resolver/map.go

```diff
@@ -136,3 +136,134 @@ func (a *AddressMap) Values() []any {
 	}
 	return ret
 }
+
+type endpointNodeList []endpointNode
+
+type endpointNode struct {
+	addrs map[string]int
+	val   any
+}
+
+// Equal returns whether the unordered set of addrs counts are the same between
+// the endpoint nodes.
+func (en endpointNode) Equal(en2 endpointNode) bool {
+	if len(en.addrs) != len(en2.addrs) {
+		return false
+	}
+	for addr, count := range en.addrs {
+		if count2, ok := en2.addrs[addr]; !ok || count != count2 {
+			return false
+		}
+	}
+	return true
+}
+
+func toEndpointNode(endpoint Endpoint) endpointNode {
+	en := make(map[string]int)
+	for _, addr := range endpoint.Addresses {
+		if count, ok := en[addr.Addr]; ok {
+			en[addr.Addr] = count + 1
+		}
+		en[addr.Addr] = 1
+	}
+	return endpointNode{
+		addrs: en,
+	}
+}
+
+// EndpointMap is a map of endpoints to arbitrary values keyed on only the
+// unordered set of addresses string within an endpoint. The reason this uses a
+// slice instead of a map is because slices or maps cannot be used as map keys.
+// Thus, for efficiency store as a map and compare to others so each node
+// comparison is o(n), where n is the number of unique addresses in the endpoint
+// rather than having to sort a list on each comparison which is n(log(n))
+// (cannot use slices for the latter case or maps for the former case as map
+// keys, so needs to be a list).
+type EndpointMap struct {
+	endpoints endpointNodeList // Cannot use maps as a map key. Thus, keep a list, and o(n) search. Not on RPC fast path so efficiency not as important.
+}
+
+// NewEndpointMap returns a new EndpointMap.
+func NewEndpointMap() *EndpointMap {
+	return &EndpointMap{}
+}
```

Comment:
Generally if the zero value of the struct is safe to use, we'd just avoid having a constructor and say the zero value is safe to use as an empty map.  E.g. https://pkg.go.dev/bytes@go1.21.1#Buffer

The exception is if you think it might need an initialization later, especially if it's an API that we can't ever change.

Reply:
As per comment below, chose not to go map route so deleted this constructor. If we do need to change this, it's internal and we won't break users :).

Reply:
This is not internal.  This is in the `grpc/resolver` package, which is experimental and intended to be accessed by users, but subject to change. But breaking changes will still break users.

Reply:
Ah, ok sounds good.